src/parsing/Parsing_common.ml

```diff
@@ -0,0 +1,384 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+open Common
+module R = Rule
+module G = AST_generic
+module J = JSON
+
+let logger = Logging.get_logger [ __MODULE__ ]
+
+(*****************************************************************************)
+(* Types *)
+(*****************************************************************************)
+
+type key = string R.wrap
+
+type env = {
+  (* id of the current rule (needed by some exns) *)
+  id : Rule_ID.t;
+  (* languages of the current rule (needed by parse_pattern) *)
+  languages : Rule.languages;
+  (* whether we are underneath a `metavariable-pattern` *)
+  in_metavariable_pattern : bool;
+  (* emma: save the path within the yaml file for each pattern
+   * (this will allow us to later report errors in playground basic mode)
+   *)
+  path : string list;
+  (* for producing decent error messages when checking the validity
+     of the options section that was parsed as a whole from JSON *)
+  options_key : key option;
+  options : Rule_options_t.t option;
+}
+
+(* Parsing generic dictionaries creates a mutable Hashtbl and consumes the
+ * fields as they are processed.
+ * todo? use a Map instead?
+ *)
+type dict = {
+  (* !this is mutated! *)
+  h : (string, key * AST_generic.expr) Hashtbl.t;
+  (* for error reports on missing fields *)
+  first_tok : R.tok;
+}
+
+(*****************************************************************************)
+(* Error Management *)
+(*****************************************************************************)
+
+let yaml_error t s = Rule.raise_error None (InvalidYaml (s, t))
+
+let yaml_error_at_expr (e : G.expr) s =
+  yaml_error (AST_generic_helpers.first_info_of_any (G.E e)) s
+
+let yaml_error_at_key (key : key) s = yaml_error (snd key) s
+
+let error rule_id t s =
+  Rule.raise_error (Some rule_id) (InvalidRule (InvalidOther s, rule_id, t))
+
+let error_at_key rule_id (key : key) s = error rule_id (snd key) s
+
+let error_at_opt_key rule_id (opt_key : key option) s =
+  match opt_key with
+  | Some key -> error_at_key rule_id key s
+  | None -> failwith (spf "Internal error (no location provided): %s" s)
+
+let error_at_expr rule_id (e : G.expr) s =
+  error rule_id (AST_generic_helpers.first_info_of_any (G.E e)) s
+
+let pcre_error_to_string s exn =
+  let message =
+    match exn with
+    | Pcre.Partial -> "String only matched the pattern partially"
+    | BadPartial ->
+        "Pattern contains items that cannot be used together with partial \
+         matching."
+    | BadPattern (msg, pos) -> spf "%s at position %d" msg pos
+    | BadUTF8 -> "UTF8 string being matched is invalid"
+    | BadUTF8Offset ->
+        "Gets raised when a UTF8 string being matched with offset is invalid."
+    | MatchLimit ->
+        "Maximum allowed number of match attempts with\n\
+        \                      backtracking or recursion is reached during \
+         matching."
```

Comment:
Extra `\` here?

Reply:
should be copied verbatim from the source file

Reply:
Hmm. It is, just seems off.